internal/releaseapi/client.go

```diff
@@ -87,48 +91,101 @@ func (c *Client) ListReleases() (ReleaseIndex, error) {
 
 func (c *Client) DownloadRelease(r Release, os, arch string) (string, error) {
 	var matchingBuild Build
-
+	var checkSha256Sum string
 	for _, build := range r.Builds {
 		if build.OS == os && build.Arch == arch {
 			matchingBuild = build
 			break
 		}
 	}
 
+	checkSums, err := c.getReleaseCheckSums(r)
+	if err != nil {
+		return "", errors.Wrap(err, "could not download checksum file")
+	}
+
+	for _, line := range strings.Split(checkSums, "\n") {
+		checksum := strings.Split(line, "  ")
+		if checksum[1] == matchingBuild.zipFileName() {
+			checkSha256Sum = checksum[0]
+			break
+		}
+	}
+
 	if matchingBuild.URL == "" {
 		return "", errors.Errorf(
 			"could not find matching build for OS '%s' and arch '%s'", os, arch,
 		)
 	}
 
-	return c.downloadBuild(matchingBuild)
+	build, sha256sum, err := c.downloadBuild(matchingBuild)
+	if sha256sum != nil { // new download
+		if checkSha256Sum != hex.EncodeToString(sha256sum) {
+			return "", errors.Errorf(
+				"checksum for %s should be %s, got %s", matchingBuild.URL, checkSha256Sum, hex.EncodeToString(sha256sum),
+			)
+		} else {
+			log.Printf("check sum match\n")
+		}
+	}
+
+	return build, err
 }
 
-func (c *Client) downloadBuild(build Build) (string, error) {
+func (c *Client) getReleaseCheckSums(release Release) (string, error) {
+	shaSumFile, _, err := c.downloadReleaseArchive(release.ShaSumsURL())
+	defer os.Remove(shaSumFile.Name())
+	defer shaSumFile.Close()
+	if err != nil {
+		return "", errors.Wrap(err, "could not download checksum file")
+	}
+	checkSums, err := os.ReadFile(shaSumFile.Name())
+	return string(checkSums), err
+}
+
+func (r *Release) ShaSumsURL() string {
+	releaseUrl := "https://releases.hashicorp.com/terraform"
+	return fmt.Sprintf("%s/%s/%s", releaseUrl, r.Version, r.Shasums)
+}
+
+func (c *Client) downloadBuild(build Build) (string, []byte, error) {
```

Comment:
I think I'd like to see the checksum _passed in_ to this function, rather than returned - `downloadBuild` can verify the checksum as a part of the download, and return an error if it doesn't match.